src/librustc_mir/hair/pattern/mod.rs

```diff
@@ -91,6 +91,7 @@ pub enum PatternKind<'tcx> {
     AscribeUserType {
         user_ty: PatternTypeProjection<'tcx>,
         subpattern: Pattern<'tcx>,
+        variance: ty::Variance,
```

Comment:
Here we should add a comment:

Variance is to use when relating the type `user_ty` to the **type of the value being matched**. Typically, this is `Covariant`, since the value being matched must have a type that is some subtype of the ascribed type. 

Note that this variance does not apply for any bindings within `subpattern`. The type assigned to those bindings must be exactly equal to the `user_ty` given here.

The one place where this field is not `Covariant` is when matching constants, where we currently use `Contravariant` -- this is because the constant type just needs to be "comparable" to the type of the input value. So for example `match x { "foo" => .. }` requires that `&'static str <: T_x`, where `T_x` is the type of `x`. Really, we should probably be checking for a `PartialEq` impl instead, but this preserves the behavior of the old type-check for now. See #57280 for details.

Reply:
Added.